### ISSUE TITLE
Fail or warn when using an overrestricitve filter that filters everything

### ISSUE THREAD

**Issue by phansch** [rust-lang/cargo#6151]
I'm not sure if this or the rust repo (libtest) is the right place to report this.

It's possible to run `cargo test <pattern>` to run the tests that match the given pattern.

However, if I run `cargo test this_does_not_exist`, everything looks green in the output. For example, if I run `cargo test mod_floors` in [Chrono](https://github.com/chronotope/chrono), it will look like this:

![selection_008](https://user-images.githubusercontent.com/2042399/46580356-0fa1f480-ca1b-11e8-88fc-87e4283e15d1.png)

On the first glance, it looks like everything is green and all tests passed.
It doesn't tell me that no tests were executed. I really wanted to run `cargo test mod_floor`:

![selection_009](https://user-images.githubusercontent.com/2042399/46580368-598ada80-ca1b-11e8-8d83-68535c2bc5d6.png)

It would be nice if `cargo test` somehow tells me that no tests were executed.

Maybe adding a yellow or red warning text could be a good start?

**Comment by phansch** [comment]
I see, thank you! fwiw, there is also this RFC which was never merged: https://github.com/rust-lang/rfcs/pull/1284

**Comment by dwijnand** [comment]
Maybe libtest can return a different return code when no test ran, and then cargo test can handle the combinations?

Say `2` means "no test ran".

Return codes `0` `2` mean at least one test ran and was successful, so "pass".
Return codes `1` `2` mean at least one test ran and was unsucesssful, so "fail".
Return codes `2` `2` mean not tests ran, so "fail".

WDYT?

**Comment by dwijnand** [comment]
Failure's in the eye of the beholder?

If libtest is asked to run all tests that contain "flippity-bippity", for which there are no test, "success" is.. debatable, no?

**Comment by dwijnand** [comment]
Btw, I'm kind of arguing on the topic in a vacuum. I can understand the desire to maintain the status quo, at this point.

**Comment by alexcrichton** [comment]
I agree that failing if no tests run makes the most sense if this is all we're considering, but there's far more to test suites than just the filter applied to it, so do think, yeah, that it's not globally the right thing to do to exit with a nonzero code if no tests are run.

**Comment by epage** [comment]
For #5609, I want to change the contract between libtest and cargo which would allow us to better handle cases like this

**Comment by Goirad** [comment]
Can we do something like `--fail-if-noop` or equivalent? If that's acceptable (up to naming or maybe making it an unstable flag) I'd be willing to put in a PR for it

**Comment by epage** [comment]
Until we change the contract and allow more control in cargo, this lives in libtest and libtest is in a soft feature freeze.
